tarantool/dbapi.py

```diff
@@ -0,0 +1,307 @@
+        indicated by the size parameter. If this is not possible due to the
+        specified number of rows not being available, fewer rows may be
+        returned.
+
+        An Error (or subclass) exception is raised if the previous call to
+        .execute*() did not produce any result set or no call was issued yet.
+
+        Note there are performance considerations involved with the size
+        parameter. For optimal performance, it is usually best to use the
+        .arraysize attribute. If the size parameter is used, then it is best
+        for it to retain the same value from one .fetchmany() call to the next.
+        """
+        size = size or self.arraysize
+
+        if self.rows is None:
+            raise self._c.ProgrammingError('Nothing to fetch')
+
+        if len(self.rows) < size:
+            items = self.rows
+            self.rows = []
+        else:
+            items, self.rows = self.rows[:size], self.rows[size:]
+
+        return items if len(items) else []
+
+    def fetchall(self):
+        """Fetch all (remaining) rows of a query result, returning them as a
+        sequence of sequences (e.g. a list of tuples). Note that the cursor's
+        arraysize attribute can affect the performance of this operation.
+
+        An Error (or subclass) exception is raised if the previous call to
+        .execute*() did not produce any result set or no call was issued yet.
+        """
+        if self.rows is None:
+            raise self._c.ProgrammingError('Nothing to fetch')
+
+        items = self.rows[:]
+        self.rows = []
+        return items
+
+    def setinputsizes(self, sizes):
+        """This can be used before a call to .execute*() to predefine memory
+        areas for the operation's parameters.
+        sizes is specified as a sequence - one item for each input parameter.
+        The item should be a Type Object that corresponds to the input that
+        will be used, or it should be an integer specifying the maximum
+        length of a string parameter. If the item is None, then no predefined
+        memory area will be reserved for that column (this is useful to avoid
+        predefined areas for large inputs).
+
+        This method would be used before the .execute*() method is invoked.
+
+        Implementations are free to have this method do nothing and users are
+        free to not use it."""
+
+    def setoutputsize(self, size, column=None):
+        """Set a column buffer size for fetches of large columns (e.g. LONGs,
+        BLOBs, etc.). The column is specified as an index into the result
+        sequence. Not specifying the column will set the default size for all
+        large columns in the cursor.
+        This method would be used before the .execute*() method is invoked.
+        Implementations are free to have this method do nothing and users are
+        free to not use it."""
+
+
+class Connection(BaseConnection):
+    _cursor = None
+    paramstyle = 'format'
```

Comment:
Can we stay on the `qmark` param style for this release?
It is quite easy to implement a wrapper for Cursor which would convert query to `format` param style. And this case perfectly fits to Django like this:
```
class CursorWrapper:
    """Tarantool supports only qmark and named param styles
    the easies way to send SQL query with proper for tarantool
    parameters is to convert them to qmark style.
    We need to overwrite execute and executemany methods to do it.
    """
    def __init__(self, cursor):
        self.cursor = cursor

    def execute(self, query, params=()):
        query = self.convert_query(query, len(params)) if params else query
        return self.cursor.execute(query, params)

    def executemany(self, query, param_list):
        query = self.convert_query(query, len(param_list[0])) if param_list else query
        return self.cursor.executemany(query, param_list)

    def convert_query(self, query, num_params):
        return query % tuple("?" * num_params)

    def __getattr__(self, attr):
        return getattr(self.cursor, attr)

    def __iter__(self):
        return iter(self.cursor)
```

Reply:
We discussed this briefly with Artem and it seems that ideal solution would be the following:

* Support both `qmark` and `named`.
* Declare `named` in `paramstyle`.

Reply:
Set paramstyle to named by default. Support both in this part of code:
```
        if isinstance(args, dict):
            args = tuple(args.values())
        try:
            request_body = msgpack.dumps({IPROTO_SQL_TEXT: sql,
                                          IPROTO_SQL_BIND: args})
        except ValueError as e:
            raise DatabaseError("Value error: %s" % e)
```

Reply:
This way it will send named parameters as a list: so they will be bind in some arbitrary order. Our documentation about the binary protocol does not shed light how to do so, so I filed the [issue](https://github.com/tarantool/doc/issues/1481).

Reply:
This format of parameter worked out:
[{":%s" % name: value}]
parameter key need to have leading semicolon